lib/pos_mempool.go

```diff
@@ -0,0 +1,446 @@
+	}
+
+	if !timeBucketNext {
+		// If there are no more transactions in the current TimeBucket, we will advance the feeBucketIterator.
+		if !fti.feeBucketIterator.Next() {
+			// If there are no more TimeBuckets in the FeeBucket, we are done.
+			return false
+		}
+
+		// If there are more TimeBuckets in the FeeBucket, we will advance the timeBucketIterator to the first transaction
+		// in the new TimeBucket.
+		if timeBucketVal, ok := fti.feeBucketIterator.Value().(*TimeBucket); ok {
+			fti.timeBucketIterator = timeBucketVal.txnsSet.Iterator()
+			// We will advance the timeBucketIterator to the first transaction in the new TimeBucket.
+			if !fti.timeBucketIterator.Next() {
+				// This shouldn't happen, but if it does, we've reached the end of the FeeBucket.
+				return false
+			}
+		} else {
+			return false
+		}
+	}
+	return true
+}
+
+// Value returns the transaction that the iterator is pointing at.
+func (fti *FeeTimeIterator) Value() (*MempoolTx, bool) {
+	if !fti.initialized {
+		return nil, false
+	}
+
+	if txn, ok := fti.timeBucketIterator.Value().(*MempoolTx); ok {
+		return txn, true
+	} else {
+		return nil, false
+	}
+}
+
+// ========================
+//	TimeBucket
+// ========================
+
+// TimeBucket is a data structure storing MempoolTx with similar fee rates. It has an index, which determines the
+// range of fee rates that are accepted by the TimeBucket.
+type TimeBucket struct {
+	// feeMin is the minimum fee rate (inclusive) accepted by the TimeBucket, in nanos per KB.
+	feeMin uint64
+	// feeMax is the maximum fee rate (inclusive) accepted by the TimeBucket, in nanos per KB.
+	feeMax uint64
+	// index is the fee range index of the TimeBucket.
+	index int
+	// txnsSet is a set of MempoolTx transactions stored in the TimeBucket.
+	txnsSet *treeset.Set
+
+	globalParams *GlobalParamsEntry
+}
+
+func NewTimeBucket(index int, globalParams *GlobalParamsEntry) *TimeBucket {
+	txnsSet := treeset.NewWith(TimeBucketComparator)
+
+	feeMin, feeMax := ComputeTimeBucketRangeFromIndex(index, globalParams)
+	return &TimeBucket{
+		feeMin:       feeMin,
+		feeMax:       feeMax,
+		index:        index,
+		txnsSet:      txnsSet,
+		globalParams: globalParams,
+	}
+}
+
+// TimeBucketComparator is a comparator function for MempoolTx transactions stored inside a TimeBucket. The comparator
+// orders the transactions by smallest timestamp. In case of a tie, transactions are ordered by greatest fee rate. Finally,
+// in case of another tie, transactions are ordered by their hash.
+func TimeBucketComparator(a, b interface{}) int {
+	aVal := a.(*MempoolTx)
+	bVal := b.(*MempoolTx)
+
+	if aVal.Added.UnixNano() > bVal.Added.UnixNano() {
+		return 1
+	} else if aVal.Added.UnixNano() < bVal.Added.UnixNano() {
+		return -1
+	} else {
+		if aVal.FeePerKB < bVal.FeePerKB {
+			return 1
+		} else if aVal.FeePerKB > bVal.FeePerKB {
+			return -1
+		} else {
+			return HashToBigint(aVal.Hash).Cmp(HashToBigint(bVal.Hash))
+		}
+	}
```

Comment:
No need to nest the if blocks. if the first 2 conditions fails, then it's guaranteed for all remaining cases that the two txns have the same timestamp.
```suggestion
	} else if aVal.FeePerKB < bVal.FeePerKB {
		return 1
	} else if aVal.FeePerKB > bVal.FeePerKB {
		return -1
	} else {
		return HashToBigint(aVal.Hash).Cmp(HashToBigint(bVal.Hash))
	}
```